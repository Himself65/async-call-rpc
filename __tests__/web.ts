class Exception extends Error {
    constructor(public message: string, public name: string) {
        super(message)
    }
}
test('DOMException', async () => {
    const { ErrorResponse } = await import('../src/utils/jsonrpc')
    const { DOMException, DOMExceptionHeader, RecoverError } = await import('../src/utils/error')
    expect(DOMException).not.toBeUndefined()
    const e = RecoverError(DOMExceptionHeader + 'MyError', 'Message', 0, '')
    expect(e).toMatchSnapshot()
    expect(e).toBeInstanceOf(Exception)

    expect(ErrorResponse('id', 0, 'msg', 'stack', new Exception('msg2', 'name'))).toMatchSnapshot()
})

test('preservePauseOnException', async () => {
    const { preservePauseOnException } = await import('../src/utils/preservePauseOnException')

    const e = () => {}
    await expect(preservePauseOnException(e, async () => 1, [])).resolves.toBe(1)
    // const throws = async () => {
    //     // Jest cannot let us test unhandledRejection well https://github.com/facebook/jest/issues/5620
    //     throw new Error('unhandledRejection test')
    // }
    // await expect(preservePauseOnException(e, throws, [])).rejects.toMatchInlineSnapshot()
}, 200)

beforeAll(() => {
    globalThis.DOMException = Exception as any
    globalThis.document = {
        // @ts-ignore
        createElement() {
            globalThis[Symbol.for('unhandledRejection')] = (e) => {
                _.contentWindow.onunhandledrejection({ reason: e } as any)
            }
            const _ = {
                style: {},
                click: () => _.onclick(),
                remove: () => {},
                onclick: () => {},
                contentDocument: globalThis.document,
                contentWindow: globalThis as typeof window,
            }
            return _
        },
        // @ts-ignore
        body: { appendChild() {} },
    }
})
