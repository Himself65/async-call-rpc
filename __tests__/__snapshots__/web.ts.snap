// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOMException 1`] = `[MyError: Message]`;

exports[`DOMException 2`] = `
Object {
  "error": Object {
    "code": 0,
    "data": Object {
      "stack": "stack",
      "type": "DOMException:name",
    },
    "message": "msg",
  },
  "id": "id",
  "jsonrpc": "2.0",
}
`;
